### PR TITLE
HUB-151 Create c level and Node from p element in dsc element.

### PR DESCRIPTION
**Pull request by warmwaterkruik**
Given XML
```xml
<dsc type="combined">
  <head>Beschrijving van de series en archiefbestanddelen</head>
  <p>Let op: deze inventaris is alleen voor het inzien van de dossiers. Kijk in de index voor het zoeken op naam. <extref linktype="simple" show="new" actuate="onrequest" href="https://www.nationaalarchief.nl/onderzoeken/index/nt00446">https://www.nationaalarchief.nl/onderzoeken/index/nt00446</extref>.</p>
  <c level="file"></c>
</dsc>

```
Op dit moment wordt inderdaad de head en p onder 'dsc combined' niet meegenomen in of de description API of de tree API.
De makkelijkste oplossing zou zijn om deze informatie te veranderen tijdens het parsen in een dummy c-level zoals als is gebeurd met `<unittitle>BELANGRIJKE MEDEDELING VOOR DE ONDERZOEKER</unittitle>` in het voorbeeld hierboven. Dan weten we zeker dat het in de rest ead verwerkings-pijplijn goed gaat.

### CODE REVIEW

File: hub3/ead/ead.go [hunk 1]
```diff
@@ -233,32 +233,48 @@ func (dsc *Cdsc) NewNodeList(cfg *NodeConfig) (*NodeList, uint64, error) {
...
-	for _, cc := range dsc.Numbered {
-		node, err := NewNode(cc, []string{}, cfg)
+	nodes := make([]*Node, 0)
+	errors := make([]string, 0)
+
+	for _, p := range dsc.Cp {
+		cc := p.NewClevel()
+		node, err := NewNode(CLevel(cc), []string{}, cfg)
 		if err != nil {
-			return nil, 0, err
```
Comment (kiivihal) [inline hub3/ead/ead.go:239]: For big EADs it is important to fail early and give the error. The handling of cLevels is already fairly robust and errors that are thrown in creating of a ead.Node are fatal for proper processing.

**Review by kiivihal — CHANGES_REQUESTED** [review]
The new functionality works as expected, but there are a few things that must be changed in the implementation.

Please review my requested changes for using the node channel in the processing of cLevels and add error handling for the NewClevel function.

File: hub3/ead/ead.go [hunk 2]
```diff
@@ -233,32 +233,48 @@ func (dsc *Cdsc) NewNodeList(cfg *NodeConfig) (*NodeList, uint64, error) {
 		nl.Label = append(nl.Label, label.Head)
 	}
 
-	for _, cc := range dsc.Numbered {
-		node, err := NewNode(cc, []string{}, cfg)
+	nodes := make([]*Node, 0)
+	errors := make([]string, 0)
+
+	for _, p := range dsc.Cp {
+		cc := p.NewClevel()
```
[the remark this answers is not in the document]
Comment (warmwaterkruik) [reply hub3/ead/ead.go]: Added error handling

File: hub3/ead/parser.go [hunk 3]
```diff
@@ -359,3 +359,11 @@ func (ca *Cabstract) CleanAbstract() []string {
 func (ut *Cunittitle) Title() string {
 	return sanitizer.Sanitize(strings.TrimSpace(fmt.Sprintf("%s", ut.Raw)))
 }
+
+// NewClevel creates a fake c level series struct from the paragraph text.
+func (cp *Cp) NewClevel() *Cc {
+	fakeC := fmt.Sprintf(`<c level="file"><did><unittitle>%s</unittitle></did></c>`, cp.Raw)
+	cc := &Cc{}
+	_ = xml.Unmarshal([]byte(fakeC), cc)
```
[the remark this answers is not in the document]
Comment (warmwaterkruik) [reply hub3/ead/parser.go]: I thought unmarshal can't fail in this case. I am perhaps too optimistic and have added and returned the unmarshal error.

File: hub3/ead/ead.go [hunk 4]
```diff
@@ -233,32 +233,48 @@ func (dsc *Cdsc) NewNodeList(cfg *NodeConfig) (*NodeList, uint64, error) {
...
+	for _, cc := range dsc.Numbered {
+		node, err := NewNode(cc, []string{}, cfg)
+		if err != nil {
+			log.Warn().Msgf("Could not create node from numbered %v", err)
+			errors = append(errors, err.Error())
 			continue
 		}
-
-		// legacy add should not happen if there is a node channel
-		nl.Nodes = append(nl.Nodes, node)
```
[the remark this answers is not in the document]
Comment (warmwaterkruik) [reply hub3/ead/ead.go]: I put these lines back

File: hub3/ead/ead.go (hunk 1 above)
Comment (warmwaterkruik) [reply hub3/ead/ead.go:239]: Returned early on all three loops

File: hub3/ead/parser.go [hunk 5]
```diff
@@ -359,3 +359,11 @@ func (ca *Cabstract) CleanAbstract() []string {
 func (ut *Cunittitle) Title() string {
 	return sanitizer.Sanitize(strings.TrimSpace(fmt.Sprintf("%s", ut.Raw)))
 }
+
+// NewClevel creates a fake c level series struct from the paragraph text.
+func (cp *Cp) NewClevel() *Cc {
+	fakeC := fmt.Sprintf(`<c level="file"><did><unittitle>%s</unittitle></did></c>`, cp.Raw)
+	cc := &Cc{}
```
[the remark this answers is not in the document]
Comment (warmwaterkruik) [reply hub3/ead/parser.go:366]: I left this because the unmarshal needs a pointer argument
	cc := &Cc{}
	err := xml.Unmarshal([]byte(fakeC), cc)

**Review by kiivihal — CHANGES_REQUESTED** [review]
Can you update the changelog.md under added: 

- Create c level and Node from p element in dsc element. [[GH-56]](https://github.com/delving/hub3/pull/56)

**Comment by warmwaterkruik** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by warmwaterkruik** [comment]
recheckcla

**Comment by warmwaterkruik** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by warmwaterkruik** [comment]
recheckcla

**Comment by kiivihal** [comment]
recheckcla

**Pull request merged**